cpp/src/decisiontree/batched-levelalgo/kernels/quantiles.cu

```diff
@@ -20,29 +20,42 @@ namespace ML {
 namespace DT {
 
 template <typename T>
-__global__ void computeQuantilesSorted(T* quantiles,
-                                       const int n_bins,
-                                       const T* sorted_data,
-                                       const int length)
+__global__ void computeQuantilesKernel(
+  T* quantiles, int* unique_nbins, const T* sorted_data, const int max_nbins, const int n_rows)
 {
-  int tid          = threadIdx.x + blockIdx.x * blockDim.x;
-  double bin_width = static_cast<double>(length) / n_bins;
-  int index        = int(round((tid + 1) * bin_width)) - 1;
-  index            = min(max(0, index), length - 1);
-  if (tid < n_bins) { quantiles[tid] = sorted_data[index]; }
+  double bin_width = static_cast<double>(n_rows) / max_nbins;
 
+  for (int bin = threadIdx.x; bin < max_nbins; bin += blockDim.x) {
+    // get index by interpolation
+    int idx        = int(round((bin + 1) * bin_width)) - 1;
+    idx            = min(max(0, idx), n_rows - 1);
+    quantiles[bin] = sorted_data[idx];
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    // make quantiles unique, in-place
+    auto new_last = thrust::unique(thrust::device, quantiles, quantiles + max_nbins);
```

Comment:
https://forums.developer.nvidia.com/t/does-thrust-transform-invoke-dynamic-parallelism/44121/2

Reply:
oh I see, thanks for the article rory... haven't encountered such an error here...but have changed policy to `thrust::seq` to be on safer side 👍🏾